flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaEvolutionUtil.java

```diff
@@ -62,4 +72,193 @@ public static int[] createIndexMapping(
+                indexMapping[i] =
+                        dataValueIndex < 0 ? dataValueIndex : dataValueIndex + dataKeyCount + 2;
+            }
+        }
+
+        for (int i = 0; i < indexMapping.length; i++) {
+            if (indexMapping[i] != i) {
+                return indexMapping;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Create data projection from table projection.
+     *
+     * @param tableFields the fields of table
+     * @param dataFields the fields of underlying data
+     * @param tableProjection the projection of table
+     * @return the projection of data
+     */
+    public static int[][] createDataProjection(
+            List<DataField> tableFields, List<DataField> dataFields, int[][] tableProjection) {
+        List<Integer> tableFieldIdList =
+                tableFields.stream().map(DataField::id).collect(Collectors.toList());
+        List<Integer> dataFieldIdList =
+                dataFields.stream().map(DataField::id).collect(Collectors.toList());
+        return Arrays.stream(tableProjection)
+                .map(p -> Arrays.copyOf(p, p.length))
+                .peek(
+                        p -> {
+                            int fieldId = tableFieldIdList.get(p[0]);
+                            p[0] = dataFieldIdList.indexOf(fieldId);
+                        })
+                .filter(p -> p[0] >= 0)
+                .toArray(int[][]::new);
+    }
+
+    /**
+     * Create predicate list from data fields.
+     *
+     * @param tableFields the table fields
+     * @param dataFields the underlying data fields
+     * @param filters the filters
+     * @return the data filters
+     */
+    public static List<Predicate> createDataFilters(
+            List<DataField> tableFields, List<DataField> dataFields, List<Predicate> filters) {
+        if (filters == null) {
+            return null;
+        }
+
+        List<Predicate> dataFilters = new ArrayList<>(filters.size());
+        for (Predicate predicate : filters) {
+            dataFilters.add(createDataPredicate(tableFields, dataFields, predicate));
+        }
+        return dataFilters;
+    }
+
+    @Nullable
+    private static Predicate createDataPredicate(
+            List<DataField> tableFields, List<DataField> dataFields, Predicate predicate) {
+        if (predicate instanceof CompoundPredicate) {
+            CompoundPredicate compoundPredicate = (CompoundPredicate) predicate;
+            List<Predicate> children = compoundPredicate.children();
+            List<Predicate> dataChildren = new ArrayList<>(children.size());
+            for (Predicate child : children) {
+                Predicate dataPredicate = createDataPredicate(tableFields, dataFields, child);
+                if (dataPredicate != null) {
+                    dataChildren.add(dataPredicate);
+                }
+            }
+            return new CompoundPredicate(compoundPredicate.function(), dataChildren);
```

Comment:
We could think about predicates deeply.
Not found means the column is null, not means we can skip this predicate?

For example:
`f0 = 1 or f1 is null`. f1 is not found in the data file. This means we can not just return a `f0 = 1`, this will cause us to filter out the required data.
We can add cases for this.